shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -0,0 +1,275 @@
+  //   with PlatformViewsController. Re-evaluate that relationship.
+  //   However, I did find that this method is called from the Camera plugin.
+  // TODO(mattcarroll): detachFromGLContext requires API 16. Create solution for earlier APIs.
+  @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+  @Override
+  public SurfaceTextureEntry createSurfaceTexture() {
+    final SurfaceTexture surfaceTexture = new SurfaceTexture(0);
+    surfaceTexture.detachFromGLContext();
+    final SurfaceTextureRegistryEntry entry = new SurfaceTextureRegistryEntry(
+        nextTextureId.getAndIncrement(),
+        surfaceTexture
+    );
+    registerTexture(entry.id(), surfaceTexture);
+    return entry;
+  }
+
+  final class SurfaceTextureRegistryEntry implements TextureRegistry.SurfaceTextureEntry {
+    private final long id;
+    private final SurfaceTexture surfaceTexture;
+    private boolean released;
+
+    SurfaceTextureRegistryEntry(long id, SurfaceTexture surfaceTexture) {
+      this.id = id;
+      this.surfaceTexture = surfaceTexture;
+
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+        // The callback relies on being executed on the UI thread (unsynchronised read of mNativeView
+        // and also the engine code check for platform thread in Shell::OnPlatformViewMarkTextureFrameAvailable),
+        // so we explicitly pass a Handler for the current thread.
+        this.surfaceTexture.setOnFrameAvailableListener(onFrameListener, new Handler());
+      } else {
+        // Android documentation states that the listener can be called on an arbitrary thread.
+        // But in practice, versions of Android that predate the newer API will call the listener
+        // on the thread where the SurfaceTexture was constructed.
+        this.surfaceTexture.setOnFrameAvailableListener(onFrameListener);
+      }
+    }
+
+    private SurfaceTexture.OnFrameAvailableListener onFrameListener = new SurfaceTexture.OnFrameAvailableListener() {
+      @Override
+      public void onFrameAvailable(SurfaceTexture texture) {
+        if (released) {
+          // Even though we make sure to unregister the callback before releasing, as of Android O
+          // SurfaceTexture has a data race when accessing the callback, so the callback may
+          // still be called by a stale reference after released==true and mNativeView==null.
+          return;
+        }
+        markTextureFrameAvailable(id);
+      }
+    };
+
+    @Override
+    public SurfaceTexture surfaceTexture() {
+      return surfaceTexture;
+    }
+
+    @Override
+    public long id() {
+      return id;
+    }
+
+    @Override
+    public void release() {
+      if (released) {
+        return;
+      }
+      unregisterTexture(id);
+      surfaceTexture.release();
+      released = true;
+    }
+  }
+  //------ END TextureRegistry IMPLEMENTATION ----
+
+  // TODO(mattcarroll): describe the native behavior that this invokes
```

Comment:
Same comment for all of this, I think there are good chances for these TODO comments to be forgotten, so I'd rather either solving the documentation TODOs right now, or linking Github issues for each of them.

Reply:
ping